File: src/models/refresh_token.py
"""Domain model for refresh tokens."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from datetime import datetime


@dataclass(frozen=True)
class RefreshToken:
    """Opaque refresh token row (SHA-256 hash only; rotation family)."""

    token_hash: str
    family_id: str
    client_id: str
    sub: str
    scopes: frozenset[str]
    expires_at: datetime
    created_at: datetime
    revoked_at: datetime | None = None

File: src/lib/db/models/refresh-token.ts
import mongoose, { Schema, Document, Model } from 'mongoose';

export interface IRefreshToken extends Document {
  token_hash: string;
  family_id: string;
  client_id: string;
  sub: string;
  scopes: string[];
  revoked_at?: Date | null;
  created_at: Date;
}

export const RefreshTokenSchema = new Schema<IRefreshToken>(
  {
    token_hash: { type: String, required: true, unique: true, index: true },
    family_id: { type: String, required: true, index: true },
    client_id: { type: String, required: true },
    sub: { type: String, required: true, index: true },
    scopes: { type: [String], default: [] },
    revoked_at: { type: Date, default: null },
    created_at: { type: Date, default: Date.now },
  },
  { collection: 'refresh_tokens' }
);

export const RefreshToken: Model<IRefreshToken> =
  mongoose.models.RefreshToken ||
  mongoose.model<IRefreshToken>('RefreshToken', RefreshTokenSchema, 'refresh_tokens');
